### PR TITLE
11923: add migration to deduplicate (by kind and loan id) response sets

### PR DESCRIPTION
**Pull request by DevneyHamilton**

### CODE REVIEW

File: db/migrate/20210831154710_de_duplicate_response_sets.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      kinds_present = loan.response_sets.map { |r| r.kind }.uniq
+      kinds_present.each do |kind|
+        dup_responses = loan.response_sets.select { |r| r.kind == kind }
+        sorted_dups = dup_responses.sort_by { |r| [r.updated_at, r.id] }
+        last_index_to_change = sorted_dups.count - 2 # last index is num_dups-1. Don't change last index.
+        for i in (0..last_index_to_change) do
+          dup = sorted_dups[i]
+          new_kind = "#{dup.kind}-dup-#{i}"
+          log_str = "For loan #{loan.id}: Updating response set id #{dup.id} (last updated at #{dup.updated_at}) kind from #{dup.kind} to #{new_kind}"
+          say_with_time log_str do
```
Comment (smoyte) [inline db/migrate/20210831154710_de_duplicate_response_sets.rb:18]: Cool did not know about this one!

**Review by smoyte — COMMENTED** [review]
Looks great!

File: db/migrate/20210831154710_de_duplicate_response_sets.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            dup.update(kind: new_kind)
```
Comment (smoyte) [inline db/migrate/20210831154710_de_duplicate_response_sets.rb]: Oh, this should be `update!` or you could end up with silent failures.

File: db/migrate/20210831154710_de_duplicate_response_sets.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      loan_responses = loan.response_sets
+      regex = /\-dup\-[0-9]+/
+      responses_to_keep_most_recent = []
+      unless loan_responses.count < 2 # don't change last_updated on responses for loans that have only one of any kind
+        loan_responses.each do |r|
+          if r.kind.match(regex)
+            old_kind = r.kind.sub(regex, "")
+            log_str = "For loan #{loan.id}: Updating response set id #{r.id} (last updated at #{r.updated_at}) kind from #{r.kind} to #{old_kind}"
+            say_with_time log_str do
+              r.update(kind: old_kind)
```
Comment (smoyte) [inline db/migrate/20210831154710_de_duplicate_response_sets.rb]: Same here.

File: db/migrate/20210831154710_de_duplicate_response_sets.rb (hunk 2 above)
Comment (DevneyHamilton) [reply db/migrate/20210831154710_de_duplicate_response_sets.rb]: thanks!

**Pull request merged**